src/MultiMap.purs

```diff
@@ -0,0 +1,43 @@
+module MultiMap
+  ( MultiMap
+  , empty
+  , singleton
+  , insert
+  , delete
+  , lookup
+  ) where
+
+import Control.Alt (map)
+import Control.Bind ((>>=))
+import Control.Category ((>>>))
+import Data.Array as Array
+import Data.Bounded (class Ord)
+import Data.Function (($))
+import Data.Map as Map
+import Data.Maybe (Maybe)
+import Data.Monoid ((<>))
+
+-- | A specific-use MultiMap where values for the same key are stored in FIFO container
+--   implemented via Array
+newtype MultiMap (k :: Type) (v :: Type) = MultiMap (Map.Map k (Array v))
```

Comment:
`List` instead of `Array` would give better `insert` time complexity.

Reply:
I had run some benchmarks. It seems that this is not the case. The code also suggests the same asymptotics.

Reply:
Strange, strange. `<>` on `Array` is O(n) and `Cons` is O(1).

Reply:
Maybe n is the size of the first array in <>, which always will be 1 here.

Reply:
By `n` I mean `size(a) + size(b)`. It's very strange that benchmarks don't show quadratic behavior here, AFAIK copying arrays can't be done faster than `O(size(a)+size(b))`, so either the size is too small to observe the `n^2` or there's something I don't know about the implementation of `Array.concat` (maybe there's a rope structure behind the scenes?).

Either way, I assume the size of arrays is insignificant here so let's not do anything until there's a real problem.